# Overview

H2O.ai is an ML framework that provides a great, easy-to-use, interface for many high performance algorithms. It also provides a function called AutoML that automatically tries a number of different classification algorithms like default Random Forest (DRF), an Extremely Randomized Forest (XRT), a random grid of Gradient Boosting Machines (GBMs), a random grid of Deep Neural Nets, a fixed grid of GLMs. AutoML then trains two Stacked Ensemble models - one ensemble contains all the models, and the second ensemble contains just the best performing model from each algorithm class/family. 

The final prediction is made with the all models ensemble and should provide the best performance out of the box

In [1]:
import numpy as np
import pandas as pd

In [2]:
%load_ext autoreload
%autoreload 2

In [3]:
# Sklearn imports
from sklearn.preprocessing import Imputer
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.model_selection import cross_val_score, train_test_split, GridSearchCV
from sklearn.metrics import confusion_matrix, f1_score, accuracy_score
from sklearn.utils import resample

In [4]:
import helperFunctions

In [5]:
import h2o
from h2o.automl import H2OAutoML
h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321..... not found.
Attempting to start a local H2O server...
; OpenJDK 64-Bit Server VM (Zulu 8.20.0.5-win64) (build 25.121-b15, mixed mode)
  Starting server from C:\Users\sankalpg\AppData\Local\Continuum\anaconda3\envs\mlpy\lib\site-packages\h2o\backend\bin\h2o.jar
  Ice root: C:\Users\sankalpg\AppData\Local\Temp\tmp4qj02j02
  JVM stdout: C:\Users\sankalpg\AppData\Local\Temp\tmp4qj02j02\h2o_SANKALPG_started_from_python.out
  JVM stderr: C:\Users\sankalpg\AppData\Local\Temp\tmp4qj02j02\h2o_SANKALPG_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321... successful.


--------------------------  -------------------------------
H2O cluster uptime:         03 secs
H2O cluster timezone:       Asia/Kolkata
H2O data parsing timezone:  UTC
H2O cluster version:        3.18.0.8
H2O cluster version age:    4 months and 3 days !!!
H2O cluster name:           H2O_from_python_SANKALPG_2sozfu
H2O cluster total nodes:    1
H2O cluster free memory:    3.531 Gb
H2O cluster total cores:    4
H2O cluster allowed cores:  4
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://127.0.0.1:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Algos, AutoML, Core V3, Core V4
Python version:             3.6.4 final
--------------------------  -------------------------------

In [6]:
# Setting the random state for later use
random_state = 565

## Load datasets

In [7]:
X_train, y_train = helperFunctions.load_clean_encode('training.csv', delimiter=';')

In [8]:
X_valid, y_valid = helperFunctions.load_clean_encode('validation.csv', delimiter=';')

Make sure that the train and validation sets have the same columns

In [9]:
X_train, X_valid = helperFunctions.equalizeColumns(X_train, X_valid)

## H2O AutoML

In [10]:
train_h2o = h2o.H2OFrame(pd.concat([X_train, y_train], axis=1))
valid_h2o = h2o.H2OFrame(X_valid)

Parse progress: |█████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


In [11]:
train_h2o['classLabel'] = train_h2o['classLabel'].asfactor()

In [12]:
print(train_h2o.columns, valid_h2o.columns)

['v2', 'v3', 'v5', 'v6', 'v7', 'v10', 'v13', 'v14', 'v1_b', 'v4_u', 'v4_y', 'v8_t', 'v9_t', 'v11_t', 'v12_p', 'v12_s', 'v12_o', 'classLabel'] ['v2', 'v3', 'v5', 'v6', 'v7', 'v10', 'v13', 'v14', 'v1_b', 'v4_u', 'v4_y', 'v8_t', 'v9_t', 'v11_t', 'v12_p', 'v12_s', 'v12_o']


In [15]:
# Run AutoML for 300 seconds
aml = H2OAutoML(max_runtime_secs=300, nfolds=10)

In [16]:
aml.train(x=train_h2o.columns, y='classLabel', training_frame = train_h2o)

AutoML progress: |████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


__AutoML Statistics__

In [21]:
lb=aml.leaderboard
lb

model_id auc logloss 
 
 
 GBM_grid_0_AutoML_20180822_185603_model_19 0.994962 0.0638104 
 GBM_grid_0_AutoML_20180822_185603_model_15 0.994881 0.0736909 
 GBM_grid_0_AutoML_20180822_185603_model_20 0.994563 0.0800151 
 GBM_grid_0_AutoML_20180822_185603_model_13 0.991639 0.124533 
 StackedEnsemble_AllModels_0_AutoML_20180822_185603 0.991355 0.0349008 
 GBM_grid_0_AutoML_20180822_185603_model_26 0.989895 0.206664 
 StackedEnsemble_BestOfFamily_0_AutoML_20180822_185603 0.988795 0.0538617 
 GBM_grid_0_AutoML_20180822_185603_model_2 0.986881 0.067286 
 GBM_grid_0_AutoML_20180822_185603_model_11 0.986815 0.084471 
 GBM_grid_0_AutoML_20180822_185603_model_8 0.986737 0.0889163

In [22]:
aml.leader

Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  GBM_grid_0_AutoML_20180822_185603_model_19


ModelMetricsBinomial: gbm
** Reported on train data. **

MSE: 6.612286143144014e-06
RMSE: 0.0025714365913131155
LogLoss: 0.0009041862047124636
Mean Per-Class Error: 0.0
AUC: 1.0
Gini: 1.0
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.978078180633734: 


0     1    Error    Rate
-----  ----  ---  -------  ------------
0      2659  0    0        (0.0/2659.0)
1      0     215  0        (0.0/215.0)
Total  2659  215  0        (0.0/2874.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value    idx
---------------------------  -----------  -------  -----
max f1                       0.978078     1        158
max f2                       0.978078     1        158
max f0point5                 0.978078     1        158
max accuracy                 0.978078     1        158
max precision                0.999467     1        0
max recall                   0.978078     1        158
max specificity              0.999467     1        0
max absolute_mcc             0.978078     1        158
max min_per_class_accuracy   0.978078     1        158
max mean_per_class_accuracy  0.978078     1        158

Gains/Lift Table: Avg response rate:  7.48 %



group    cumulative_data_fraction    lower_threshold    lift     cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain     cumulative_gain
--  -------  --------------------------  -----------------  -------  -----------------  ---------------  --------------------------  --------------  -------------------------  -------  -----------------
    1        0.0100905                   0.997949           13.3674  13.3674            1                1                           0.134884        0.134884                   1236.74  1236.74
    2        0.0201809                   0.996981           13.3674  13.3674            1                1                           0.134884        0.269767                   1236.74  1236.74
    3        0.0302714                   0.995936           13.3674  13.3674            1                1                           0.134884        0.404651                   1236.74  1236.74
    4        0.0400139                   0.994713           13.3674  13.3674            1                1                           0.130233        0.534884                   1236.74  1236.74
    5        0.0501044                   0.992984           13.3674  13.3674            1                1                           0.134884        0.669767                   1236.74  1236.74
    6        0.100209                    0.00304089         6.59089  9.97917            0.493056         0.746528                    0.330233        1                          559.089  897.917
    7        0.149965                    0.00117529         0        6.66821            0                0.49884                     0               1                          -100     566.821
    8        0.20007                     0.00070376         0        4.99826            0                0.373913                    0               1                          -100     399.826
    9        0.29993                     0.000353125        0        3.33411            0                0.24942                     0               1                          -100     233.411
    10       0.400139                    0.000215367        0        2.49913            0                0.186957                    0               1                          -100     149.913
    11       0.5                         0.000138006        0        2                  0                0.149617                    0               1                          -100     100
    12       0.599861                    9.0957e-05         0        1.66705            0                0.12471                     0               1                          -100     66.7053
    13       0.70007                     5.70862e-05        0        1.42843            0                0.106859                    0               1                          -100     42.8429
    14       0.79993                     3.39647e-05        0        1.25011            0                0.0935189                   0               1                          -100     25.0109
    15       0.899791                    1.71669e-05        0        1.11137            0                0.08314                     0               1                          -100     11.1369
    16       1                           2.82007e-06        0        1                  0                0.0748086                   0               1                          -100     0



ModelMetricsBinomial: gbm
** Reported on validation data. **

MSE: 0.014314498572088228
RMSE: 0.11964321364828107
LogLoss: 0.0492739008494348
Mean Per-Class Error: 0.010754027570171076
AUC: 0.9970104633781763
Gini: 0.9940209267563527
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.05486151932175458: 


0    1    Error    Rate
-----  ---  ---  -------  -----------
0      667  2    0.003    (2.0/669.0)
1      1    53   0.0185   (1.0/54.0)
Total  668  55   0.0041   (3.0/723.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.0548615    0.972477  54
max f2                       0.0548615    0.97786   54
max f0point5                 0.0918785    0.988372  50
max accuracy                 0.0918785    0.995851  50
max precision                0.997529     1         0
max recall                   0.000778056  1         153
max specificity              0.997529     1         0
max absolute_mcc             0.0548615    0.970281  54
max min_per_class_accuracy   0.0548615    0.981481  54
max mean_per_class_accuracy  0.0548615    0.989246  54

Gains/Lift Table: Avg response rate:  7.47 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.011065                    0.9917             13.3889   13.3889            1                1                           0.148148        0.148148                   1238.89   1238.89
    2        0.0207469                   0.963659           13.3889   13.3889            1                1                           0.12963         0.277778                   1238.89   1238.89
    3        0.0304288                   0.916918           13.3889   13.3889            1                1                           0.12963         0.407407                   1238.89   1238.89
    4        0.0401107                   0.833241           13.3889   13.3889            1                1                           0.12963         0.537037                   1238.89   1238.89
    5        0.0511757                   0.621834           13.3889   13.3889            1                1                           0.148148        0.685185                   1238.89   1238.89
    6        0.100968                    0.00977037         5.95062   9.7207             0.444444         0.726027                    0.296296        0.981481                   495.062   872.07
    7        0.150761                    0.0021819          0         6.51019            0                0.486239                    0               0.981481                   -100      551.019
    8        0.200553                    0.000954449        0         4.89387            0                0.365517                    0               0.981481                   -100      389.387
    9        0.300138                    0.000421619        0.185957  3.3318             0.0138889        0.248848                    0.0185185       1                          -81.4043  233.18
    10       0.399723                    0.00026913         0         2.50173            0                0.186851                    0               1                          -100      150.173
    11       0.500692                    0.000169267        0         1.99724            0                0.149171                    0               1                          -100      99.7238
    12       0.600277                    0.000102947        0         1.6659             0                0.124424                    0               1                          -100      66.5899
    13       0.699862                    5.61832e-05        0         1.42885            0                0.106719                    0               1                          -100      42.8854
    14       0.799447                    3.5139e-05         0         1.25087            0                0.0934256                   0               1                          -100      25.0865
    15       0.899032                    1.66555e-05        0         1.11231            0                0.0830769                   0               1                          -100      11.2308
    16       1                           3.67162e-06        0         1                  0                0.0746888                   0               1                          -100      0



ModelMetricsBinomial: gbm
** Reported on cross-validation data. **

MSE: 0.017217559004664148
RMSE: 0.13121569648736445
LogLoss: 0.06381041174794941
Mean Per-Class Error: 0.03082729125305017
AUC: 0.9949622606855173
Gini: 0.9899245213710346
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.10430101146906881: 


0     1    Error    Rate
-----  ----  ---  -------  -------------
0      2651  8    0.003    (8.0/2659.0)
1      27    188  0.1256   (27.0/215.0)
Total  2678  196  0.0122   (35.0/2874.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.104301     0.914842  183
max f2                       0.0271962    0.924062  219
max f0point5                 0.18724      0.947817  168
max accuracy                 0.112291     0.987822  181
max precision                0.999932     1         0
max recall                   0.000247494  1         379
max specificity              0.999932     1         0
max absolute_mcc             0.104301     0.909426  183
max min_per_class_accuracy   0.00824808   0.966905  258
max mean_per_class_accuracy  0.0107527    0.969173  246

Gains/Lift Table: Avg response rate:  7.48 %



group    cumulative_data_fraction    lower_threshold    lift       cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  ---------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0100905                   0.990448           13.3674    13.3674            1                1                           0.134884        0.134884                   1236.74   1236.74
    2        0.0201809                   0.973523           13.3674    13.3674            1                1                           0.134884        0.269767                   1236.74   1236.74
    3        0.0302714                   0.929592           13.3674    13.3674            1                1                           0.134884        0.404651                   1236.74   1236.74
    4        0.0400139                   0.845819           13.3674    13.3674            1                1                           0.130233        0.534884                   1236.74   1236.74
    5        0.0501044                   0.620375           13.3674    13.3674            1                1                           0.134884        0.669767                   1236.74   1236.74
    6        0.100209                    0.00881828         5.84826    9.60785            0.4375           0.71875                     0.293023        0.962791                   484.826   860.785
    7        0.149965                    0.003032           0.467393   6.57517            0.034965         0.491879                    0.0232558       0.986047                   -53.2607  557.517
    8        0.20007                     0.00156435         0.185659   4.97501            0.0138889        0.372174                    0.00930233      0.995349                   -81.4341  397.501
    9        0.29993                     0.00072535         0          3.3186             0                0.24826                     0               0.995349                   -100      231.86
    10       0.400139                    0.000385703        0          2.48751            0                0.186087                    0               0.995349                   -100      148.751
    11       0.5                         0.00021298         0.0465765  2                  0.00348432       0.149617                    0.00465116      1                          -95.3424  100
    12       0.599861                    0.000119352        0          1.66705            0                0.12471                     0               1                          -100      66.7053
    13       0.70007                     6.88855e-05        0          1.42843            0                0.106859                    0               1                          -100      42.8429
    14       0.79993                     3.24472e-05        0          1.25011            0                0.0935189                   0               1                          -100      25.0109
    15       0.899791                    9.02469e-06        0          1.11137            0                0.08314                     0               1                          -100      11.1369
    16       1                           2.92088e-08        0          1                  0                0.0748086                   0               1                          -100      0


Cross-Validation Metrics Summary: 


mean       sd          cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid    cv_6_valid    cv_7_valid    cv_8_valid    cv_9_valid    cv_10_valid
-----------------------  ---------  ----------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  -------------
accuracy                 0.989905   0.00280225  0.993056      0.993056      0.989583      0.996528      0.989547      0.989547      0.986063      0.993031      0.982578      0.986063
auc                      0.993376   0.00684402  0.999316      0.999287      0.965653      0.999658      0.995525      0.99554       0.98988       0.998926      0.994271      0.995703
err                      0.0100949  0.00280225  0.00694444    0.00694444    0.0104167     0.00347222    0.010453      0.010453      0.0139373     0.00696864    0.0174216     0.0139373
err_count                2.9        0.803119    2             2             3             1             3             3             4             2             5             4
f0point5                 0.943337   0.0196735   0.954545      0.952381      0.943396      0.990566      0.940594      0.969388      0.957447      0.929203      0.891089      0.904762
f1                       0.930492   0.0201798   0.954545      0.952381      0.930233      0.976744      0.926829      0.926829      0.9           0.954545      0.878049      0.904762
f2                       0.918948   0.029322    0.954545      0.952381      0.917431      0.963303      0.913462      0.88785       0.849057      0.981308      0.865385      0.904762
lift_top_group           13.3745    0.213859    13.0909       13.7143       13.0909       13.0909       13.6667       13.0455       13.0455       13.6667       13.6667       13.6667
logloss                  0.063832   0.0171153   0.0484987     0.0278088     0.0884303     0.0286577     0.0823394     0.0556475     0.0856369     0.0446571     0.0922864     0.0843568
max_per_class_error      0.0888471  0.0360447   0.0454545     0.0476191     0.0909091     0.0454545     0.0952381     0.136364      0.181818      0.0075188     0.142857      0.0952381
mcc                      0.926106   0.0213513   0.950786      0.948636      0.924899      0.975177      0.921522      0.924104      0.897784      0.951934      0.868974      0.897243
mean_per_class_accuracy  0.954073   0.018166    0.975393      0.974318      0.952666      0.977273      0.950501      0.931818      0.909091      0.996241      0.924812      0.948622
mean_per_class_error     0.0459266  0.018166    0.024607      0.0256822     0.0473342     0.0227273     0.0494987     0.0681818     0.0909091     0.0037594     0.075188      0.0513784
mse                      0.0172229  0.00430423  0.0160153     0.00862021    0.0207275     0.00803889    0.0218484     0.0128332     0.0233164     0.0125948     0.0255581     0.0226766
precision                0.952711   0.0258561   0.954545      0.952381      0.952381      1             0.95          1             1             0.913044      0.9           0.904762
r2                       0.750789   0.0630623   0.773005      0.872482      0.706216      0.88606       0.677832      0.818687      0.670574      0.814281      0.62313       0.66562
recall                   0.911905   0.0369176   0.954545      0.952381      0.909091      0.954545      0.904762      0.863636      0.818182      1             0.857143      0.904762
rmse                     0.12895    0.0172448   0.126552      0.0928451     0.14397       0.0896598     0.147812      0.113283      0.152697      0.112227      0.159869      0.150588
specificity              0.996242   0.00205911  0.996241      0.996255      0.996241      1             0.996241      1             1             0.992481      0.992481      0.992481

Scoring History: 


timestamp            duration          number_of_trees    training_rmse          training_logloss       training_auc        training_lift       training_classification_error    validation_rmse      validation_logloss    validation_auc      validation_lift    validation_classification_error
---  -------------------  ----------------  -----------------  ---------------------  ---------------------  ------------------  ------------------  -------------------------------  -------------------  --------------------  ------------------  -----------------  ---------------------------------
     2018-08-22 19:00:07  3 min 19.135 sec  0.0                0.2630823029059509     0.2659034168955254     0.5                 1.0                 0.9251913709116214               0.26288855945194894  0.2656020303174176    0.5                 1.0                0.9253112033195021
     2018-08-22 19:00:07  3 min 19.151 sec  5.0                0.21059398206441712    0.1629370888058171     0.9937561769156091  13.367441860465116  0.017745302713987474             0.22973631335989614  0.1910847507972272    0.9603471184188672  13.38888888888889  0.04979253112033195
     2018-08-22 19:00:07  3 min 19.175 sec  10.0               0.16881303939705963    0.11459248001237746    0.9995661946701418  13.367441860465116  0.004523312456506611             0.2033729283096886   0.1497541146139771    0.9836405912639097  13.38888888888889  0.02074688796680498
     2018-08-22 19:00:07  3 min 19.199 sec  15.0               0.13685322916971465    0.08432449267962805    0.9999064169953734  13.367441860465116  0.0024356297842727907            0.18658641909752732  0.12651830785938417   0.9834468250013841  13.38888888888889  0.019363762102351315
     2018-08-22 19:00:07  3 min 19.227 sec  20.0               0.11327563697208201    0.06440200574145508    0.9999825078496025  13.367441860465116  0.0010438413361169101            0.17809817224885297  0.11316091774981978   0.9867685323589659  13.38888888888889  0.02351313969571231
---  ---                  ---               ---                ---                    ---                    ---                 ---                 ---                              ---                  ---                   ---                 ---                ---
     2018-08-22 19:00:08  3 min 19.583 sec  90.0               0.005648192220738084   0.002054878856723183   1.0                 13.367441860465116  0.0                              0.12344850973798081  0.05143674117606808   0.9968582184576206  13.38888888888889  0.004149377593360996
     2018-08-22 19:00:08  3 min 19.603 sec  95.0               0.0043782407828437215  0.001599694777064786   1.0                 13.367441860465116  0.0                              0.12347072418651397  0.051986393638640396  0.9967474948790344  13.38888888888889  0.005532503457814661
     2018-08-22 19:00:08  3 min 19.631 sec  100.0              0.003449165829884665   0.0012324716665097124  1.0                 13.367441860465116  0.0                              0.12142938829410517  0.050467229504386965  0.9968997397995905  13.38888888888889  0.005532503457814661
     2018-08-22 19:00:08  3 min 19.663 sec  105.0              0.002840151921936756   0.0009935739142859306  1.0                 13.367441860465116  0.0                              0.11958251782919263  0.048950548394842676  0.9971765487460554  13.38888888888889  0.004149377593360996
     2018-08-22 19:00:08  3 min 19.679 sec  107.0              0.0025714365913131155  0.0009041862047124636  1.0                 13.367441860465116  0.0                              0.11964321364828107  0.0492739008494348    0.9970104633781763  13.38888888888889  0.004149377593360996


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
v6          102.049                1                    0.153739
v3          100.501                0.984827             0.151406
v8_t        96.8396                0.94895              0.14589
v7          84.913                 0.832079             0.127923
v13         70.0271                0.686209             0.105497
v14         62.8388                0.61577              0.0946677
v2          35.1191                0.344139             0.0529075
v10         30.7486                0.301312             0.0463233
v5          24.4349                0.239442             0.0368116
v9_t        21.2211                0.20795              0.03197
v11_t       8.69666                0.0852203            0.0131017
v4_u        8.34688                0.0817927            0.0125747
v1_b        6.40559                0.0627696            0.00965013
v4_y        6.17808                0.0605402            0.00930738
v12_s       5.4525                 0.0534301            0.00821427
v12_p       0.0109522              0.000107323          1.64997e-05

__Validation Performance__

In [23]:
preds = aml.predict(valid_h2o)

Parse progress: |█████████████████████████████████████████████████████████| 100%
gbm prediction progress: |████████████████████████████████████████████████| 100%


In [24]:
print(valid_h2o.shape, preds.shape)

(195, 17) (195, 3)


In [43]:
accuracy_score(y_pred=preds['predict'].as_data_frame().values, y_true=y_valid)

0.8666666666666667

# Summary

The best performance with the AutoML classifier was about 86.67% accuracy on the validation dataset which is the highest performance seen in this exercise.
